Framework/PythonInterface/mantid/geometry/src/Exports/ComponentInfoPythonIterator.cpp

```diff
@@ -0,0 +1,41 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+#include "MantidPythonInterface/api/ComponentInfoPythonIterator.h"
+
+#include <boost/python/class.hpp>
+#include <boost/python/copy_const_reference.hpp>
+#include <boost/python/iterator.hpp>
+#include <boost/python/module.hpp>
+
+using Mantid::PythonInterface::ComponentInfoPythonIterator;
+using namespace boost::python;
+
+// Export ComponentInfoPythonIterator
+void export_ComponentInfoPythonIterator() {
+
+  // Export to Python
+  class_<ComponentInfoPythonIterator>("ComponentInfoPythonIterator", no_init)
+      .def("__iter__", objects::identity_function())
+      .def(
+#if PY_VERSION_HEX >= 0x03000000
+          "__next__"
+#else
+          "next"
+#endif
+          ,
+          &ComponentInfoPythonIterator::next,
+          return_value_policy<copy_const_reference>());
+  /*
+   Return value policy for next is to copy the const reference. Copy by value is
+   essential for python 2.0 compatibility because items (DetectorInfoItem) will
+   outlive their iterators if declared as part of for loops. There is no good
+   way to deal with this other than to force a copy so that internals of the
+   item are not also corrupted. Future developers may wish to choose a separte
```

Comment:
What was the reason for not returning the items by value from the iterator?

Reply:
I think that again comes in with the c++ iterator with `iterator_facade` see declaration mention above.

Reply:
It defaults to returning `value_type &`, but I believe you can change that with an extra template argument.